traits/trait_list_object.py

```diff
@@ -843,12 +734,64 @@ def __setstate__(self, state):
         state["notifiers"] = [self.notifier]
         object = state.pop("object", None)
         if object is not None:
-            state['object'] = ref(object)
+            state["object"] = ref(object)
             trait = self.object()._trait(name, 0)
             if trait is not None:
-                state['trait'] = trait.handler
+                state["trait"] = trait.handler
         else:
-            state['object'] = lambda: None
-            state['trait'] = None
+            state["object"] = lambda: None
+            state["trait"] = None
 
         self.__dict__.update(state)
+
+    # -- private methods ------------------------------------------------------
+
+    def _item_validator(self, value):
+        """
+        Validate an item that's being added to the list.
+        """
+        object = self.object()
+        if object is None:
+            return value
+
+        trait_validator = self.trait.item_trait.handler.validate
+        if trait_validator is None:
+            return value
+
+        try:
+            return trait_validator(object, self.name, value)
+        except TraitError as excp:
+            excp.set_prefix("Each element of the")
+            raise
+
+    def _validate_length(self, new_length):
+        """
+        Validate the new length for a proposed operation.
+
+        Parameters
+        ----------
+        new_length : int
+            New length of the list.
+
+        Raises
+        ------
+        TraitError
+            If the proposed new length would violate the length constraints
+            of the list.
+        """
+        trait = getattr(self, "trait", None)
+        if trait is None:
```

Comment:
This code was copied from the earlier implementation, but I don't know why this check would be necessary. Ideally it should have a test that exercises this branch (or if it's not necessary, it should be removed).

Reply:
Ah, this is about pickling again, and there _are_ tests elsewhere in the test suite that exercise this case.

It's the same problem as usual: when unpickling, the sequence of events is:

1. `__new__` is called 
2. the `append` or `extend` method is called to populate the list items
3. we call the general `__setstate__` method.

But in step `2` we're relying on the `trait` attribute already being present.

Ideally we'd re-fix this in such a way that we don't need to make this check in the common path.

I'd like to re-examine the whole of the pickling structure here, but I think that's out of scope for this PR.

Reply:
Opened #996 for this. Resolving my own comments here.